rust/agama-network/src/dbus/interfaces.rs

```diff
@@ -0,0 +1,402 @@
+    ) -> zbus::fdo::Result<()> {
+        let actions = self.actions.lock();
+        actions
+            .send(Action::UpdateConnection(connection.clone()))
+            .unwrap();
+        Ok(())
+    }
+}
+
+#[dbus_interface(name = "org.opensuse.Agama.Network1.Connection.IPv4")]
+impl Ipv4 {
+    #[dbus_interface(property)]
+    pub fn addresses(&self) -> Vec<(String, u32)> {
+        let connection = self.get_connection();
+        connection
+            .ipv4()
+            .addresses
+            .iter()
+            .map(|(addr, prefix)| (addr.to_string(), *prefix))
+            .collect()
+    }
+
+    #[dbus_interface(property)]
+    pub fn set_addresses(&mut self, addresses: Vec<(String, u32)>) -> zbus::fdo::Result<()> {
+        let mut connection = self.get_connection();
+        addresses
+            .iter()
+            .map(|(addr, prefix)| addr.parse::<Ipv4Addr>().map(|a| (a, *prefix)))
+            .collect::<Result<Vec<(Ipv4Addr, u32)>, AddrParseError>>()
+            .and_then(|parsed| Ok(connection.ipv4_mut().addresses = parsed))
+            .map_err(|err| NetworkStateError::from(err))?;
+        self.update_connection(connection)
+    }
+
+    #[dbus_interface(property)]
+    pub fn method(&self) -> u8 {
+        let connection = self.get_connection();
+        connection.ipv4().method as u8
+    }
+
+    #[dbus_interface(property)]
+    pub fn set_method(&mut self, method: u8) -> zbus::fdo::Result<()> {
+        let mut connection = self.get_connection();
+        connection.ipv4_mut().method = method.try_into()?;
+        self.update_connection(connection)
+    }
+
+    #[dbus_interface(property)]
+    pub fn nameservers(&self) -> Vec<String> {
+        let connection = self.get_connection();
+        connection
+            .ipv4()
+            .nameservers
+            .iter()
+            .map(|a| a.to_string())
+            .collect()
+    }
+
+    #[dbus_interface(property)]
+    pub fn set_nameservers(&mut self, addresses: Vec<String>) -> zbus::fdo::Result<()> {
+        let mut connection = self.get_connection();
+        let ipv4 = connection.ipv4_mut();
+        addresses
+            .iter()
+            .map(|addr| addr.parse::<Ipv4Addr>())
+            .collect::<Result<Vec<Ipv4Addr>, AddrParseError>>()
+            .and_then(|parsed| Ok(ipv4.nameservers = parsed))
+            .map_err(|err| NetworkStateError::from(err))?;
+        self.update_connection(connection)
+    }
+
+    #[dbus_interface(property)]
+    pub fn gateway(&self) -> String {
+        let connection = self.get_connection();
+        match connection.ipv4().gateway {
+            Some(addr) => addr.to_string(),
+            None => "".to_string(),
+        }
+    }
+
+    #[dbus_interface(property)]
+    pub fn set_gateway(&mut self, gateway: String) -> zbus::fdo::Result<()> {
+        let mut connection = self.get_connection();
+        let ipv4 = connection.ipv4_mut();
+        gateway
+            .parse::<Ipv4Addr>()
```

Comment:
if I get code, it means you cannot set `""` to clear gateway. Am I right? If so, how to unset it?

Reply:
Good point. I have fixed it.